plugin/plugins/filesystem/nodes/utils.go

```diff
@@ -0,0 +1,278 @@
+package fsnodes
+
+import (
+	"context"
+	"hash/fnv"
+	"time"
+
+	"github.com/djdv/p9/p9"
+	"github.com/ipfs/go-cid"
+	ipld "github.com/ipfs/go-ipld-format"
+	logging "github.com/ipfs/go-log"
+	"github.com/ipfs/go-unixfs"
+	unixpb "github.com/ipfs/go-unixfs/pb"
+	coreiface "github.com/ipfs/interface-go-ipfs-core"
+	corepath "github.com/ipfs/interface-go-ipfs-core/path"
+)
+
+func doClone(names []string) bool {
+	l := len(names)
+	if l < 1 {
+		return true
+	}
+	//TODO: double check the spec to make sure dot handling is correct
+	// we may only want to clone on ".." if we're a root
+	if pc := names[0]; l == 1 && (pc == ".." || pc == "." || pc == "") {
+		return true
+	}
+	return false
+}
+
+//TODO: rename this and/or extend
+// it only does some of the stat and not what people probably expect
+func coreStat(ctx context.Context, dirEnt *p9.Dirent, core coreiface.CoreAPI, path corepath.Path) error {
+	if ipldNode, err := core.ResolveNode(ctx, path); err != nil {
+		return err
+	} else {
+		return ipldStat(dirEnt, ipldNode)
+	}
+}
+
+func coreGetAttr(ctx context.Context, attr *p9.Attr, attrMask p9.AttrMask, core coreiface.CoreAPI, path corepath.Path) (err error) {
+	ipldNode, err := core.ResolveNode(ctx, path)
+	if err != nil {
+		return err
+	}
+	ufsNode, err := unixfs.ExtractFSNode(ipldNode)
+	if err != nil {
+		return err
+	}
+
+	if attrMask.Mode {
+	attr.Mode = IRXA //TODO: this should probably be the callers responsability; just document that permissions should be set afterwards or something
+	attr.Mode |= unixfsTypeTo9Mode(ufsNode.Type())
+	}
+
+	if attrMask.Blocks{
+	if bs := ufsNode.BlockSizes(); len(bs) != 0 {
+		attr.BlockSize = bs[0] //NOTE: this value is to be used as a hint only; subsequent child block size may differ
+	}
+
+	//TODO [eventualy]: switch off here for handling of time metadata in new format standard
+	timeStamp(attr, attrMask)
+}
+
+if attrMask.Size {
+	attr.Size, attrMask.Size = ufsNode.FileSize(), true
+}
+
+if attrMask.RDev {
+	switch path.Namespace() {
+	case "ipfs":
+		attr.RDev, attrMask.RDev = dIPFS, true
+		//case "ipns":
+		//attr.RDev, attrMask.RDev = dIPNS, true
+		//etc.
+	}
+}
+
+	return nil
+}
+
+func ipldStat(dirEnt *p9.Dirent, node ipld.Node) error {
+	ufsNode, err := unixfs.ExtractFSNode(node)
+
+	if err != nil {
+		return err
+	}
+
+	nodeType := unixfsTypeTo9Mode(ufsNode.Type()).QIDType()
+
+	dirEnt.Type = nodeType
+	dirEnt.QID.Type = nodeType
+	dirEnt.QID.Path = cidToQPath(node.Cid())
+
+	return nil
+}
+
+func cidToQPath(cid cid.Cid) uint64 {
```

Comment:
This won't work as an attacker can generate CIDs such that two QIDs conflict. Do QIDs need to persist across reboots? If not, we'll have to store them in a table. If so, we're in trouble.

Also, looking at some QID documentation (http://knusbaum.com/useful/rfc9p2000), we _definitely_ wont' be able to use CIDs in the writable case.

Related question: Do p9-based _copy_ tools assume that two files with the same QID should be _hardlinks_ of each other? If that's the case, `cp -a /ipfs/QmFoo... .` will produce really weird results.

Reply:
Digging into this a bit... We're fine. plan9 doesn't even have a recursive copy tool, it recommends two invocations of tar :scream:.

Reply:
You know, this really sucks. We can't delete anything from this table, can we.

Reply:
Ok, so we _may_ be able to seed the FNV hasher such that it's secure. We'll have to work through this carefully.

Reply:
We talked about this _extensively_ on a call. Conclusions:

* Even with a random seed, we get 32 bits of security due to the birthday paradox.
* Even _fuse_ needs inodes so we aren't any worse-off using 9P.
  * However, fuse _may_ have a way to tell the OS "please forget this inode". That would allow us to use the table method.
* With a seed, a remote attacker would need us to read ~2^32 files to find a collision.
* We can (if necessary) keep an in-memory table of _open_ files to prevent _open_ files from colliding.
* We don't have to remember QIDs across restarts.
* A random collision likely isn't going to be a big deal from a security standpoint. To be effective, an attack would probably have to target a _specific_ file which would require 2^64 attempts (infeasible: 1ns to open a file -> 500 years).
* A random collision due to heavy use likely won't even be noticed. The OS likely won't care (unless the file is currently open) and checking inodes at the application layer is rare and usually limited to very specific cases. Such cases usually do not include comparing 2^32 files.

TL;DR:

1. Use a seed and regenerate it on every restart.
2. Later, keep a table of open files and in-use QIDs.
3. Even later, consider keeping an LRU of QIDs.

Reply:
Addressed (1.) in https://github.com/ipfs/go-ipfs/pull/6612/commits/e5ec67218d3a297f9f9d642251c7caae2ee6712a